hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAirdropHandler.java

```diff
@@ -0,0 +1,567 @@
+    }
+
+    /**
+     * As part of pre-handle, token transfers in the transfer list are plausible.
+     *
+     * @param tokenID      The ID of the token we are transferring
+     * @param transfers    The transfers to check
+     * @param ctx          The context we gather signing keys into
+     * @param accountStore The account store to use to look up accounts
+     * @throws PreCheckException If the transaction is invalid
+     */
+    private void checkFungibleTokenTransfers(
+            @NonNull final TokenID tokenID,
+            @NonNull final List<AccountAmount> transfers,
+            @NonNull final PreHandleContext ctx,
+            @NonNull final ReadableAccountStore accountStore)
+            throws PreCheckException {
+        final var tokenStore = ctx.createStore(ReadableTokenStore.class);
+        final var tokenRelStore = ctx.createStore(ReadableTokenRelationStore.class);
+        // Fail if we have custom fees attached to the token
+        validateTruePreCheck(tokenHasNoCustomFeesPaidByReceiver(tokenID, tokenStore), INVALID_TRANSACTION);
+        // We're going to iterate over all the transfers in the transfer list. Each transfer is known as an
+        // "account amount". Each of these represents the transfer of fungible token INTO a single account or OUT of a
+        // single account.
+        for (final var accountAmount : transfers) {
+            // Given an accountId, we need to look up the associated account.
+            final var accountId = validateAccountID(accountAmount.accountIDOrElse(AccountID.DEFAULT), null);
+            final var account = accountStore.getAliasedAccountById(accountId);
+            final var isCredit = accountAmount.amount() > 0;
+            final var isDebit = accountAmount.amount() < 0;
+            if (account != null) {
+                if (isStakingAccount(ctx.configuration(), account.accountId()) && (isDebit || isCredit)) {
+                    throw new PreCheckException(ACCOUNT_IS_IMMUTABLE);
+                }
+
+                if (isDebit) {
+                    final var tokenRel = tokenRelStore.get(accountId, tokenID);
+                    validateTruePreCheck(tokenRel != null, TOKEN_NOT_ASSOCIATED_TO_ACCOUNT);
+                    if (accountAmount.isApproval()) {
+                        final var topLevelPayer = ctx.payer();
+                        final var tokenAllowances = new ArrayList<>(account.tokenAllowances());
+                        var haveExistingAllowance = false;
+                        for (final var allowance : tokenAllowances) {
+                            if (topLevelPayer.equals(allowance.spenderId()) && tokenID.equals(allowance.tokenId())) {
+                                haveExistingAllowance = true;
+                                final var newAllowanceAmount = allowance.amount() + accountAmount.amount();
+                                validateTruePreCheck(newAllowanceAmount >= 0, AMOUNT_EXCEEDS_ALLOWANCE);
```

Comment:
Shouldn't we be able to `break` here? (I.e., this list must not contain multiple allowances with the same `(spenderId, tokenId)` pair.)

Reply:
added break